File: src/main/java/net/maizegenetics/analysis/data/ExportMultiplePlugin.java
/*
 * ExportMultiplePlugin.java
 *
 * Created on December 21, 2010
 *
 */
package net.maizegenetics.analysis.data;

import net.maizegenetics.plugindef.AbstractPlugin;
import net.maizegenetics.plugindef.DataSet;
import net.maizegenetics.plugindef.Datum;
import net.maizegenetics.util.Utils;
import org.apache.log4j.Logger;

import java.io.File;
import java.util.List;

/**
 * @author Terry Casstevens
 */
public class ExportMultiplePlugin extends AbstractPlugin {

    private static final Logger myLogger = Logger.getLogger(ExportMultiplePlugin.class);
    private FileLoadPlugin.TasselFileType[] myFileTypes = null;
    private String[] mySaveFiles = null;
    private final ExportPlugin myExportPlugin;

    /**
     * Creates a new instance of ExportMultiplePlugin
     */
    public ExportMultiplePlugin() {
        super(false);
        myExportPlugin = new ExportPlugin(false);
    }

    @Override
    public DataSet performFunction(DataSet input) {

        List data = input.getDataSet();

        int numSaveFiles = 0;
        if (mySaveFiles != null) {
            numSaveFiles = mySaveFiles.length;
        }

        if ((numSaveFiles != 0) && (numSaveFiles != 1) && (numSaveFiles != data.size())) {
            throw new IllegalStateException("ExportMultiplePlugin: performFunction: number of save files should be either 0, 1 or number of input data sets.");
        }

        if ((myFileTypes != null) && (myFileTypes.length != 0)) {
            if ((myFileTypes.length != 1) && (myFileTypes.length != data.size())) {
                throw new IllegalStateException("ExportMultiplePlugin: performFunction: number of files types should be either 0, 1 or number of input data sets.");
            }
        }

        for (int i = 0, n = data.size(); i < n; i++) {

            Datum datum = (Datum) data.get(i);
            DataSet current = new DataSet(datum, input.getCreator());

            if (numSaveFiles == 0) {
                myExportPlugin.saveFile(datum.getName());
            } else if (numSaveFiles == 1) {
                String temp;
                if (data.size() == 1) {
                    temp = mySaveFiles[0];
                } else {
                    String filename = Utils.getFilename(mySaveFiles[0]);
                    temp = filename.replaceFirst("\\.", (i + 1) + ".");
                    if (temp.length() == filename.length()) {
                        temp = filename + (i + 1);
                    }
                    String directory = Utils.getDirectory(mySaveFiles[0]);
                    if (!directory.equals(".")) {
                        File dir = new File(directory);
                        if (!dir.exists()) {
                            dir.mkdirs();
                        }
                        temp = directory + "/" + temp;
                    }
                }
                myExportPlugin.saveFile(temp);
            } else {
                myExportPlugin.saveFile(mySaveFiles[i]);
            }

            if ((myFileTypes == null) || (myFileTypes.length == 0)) {
                myExportPlugin.fileType(null);
            } else if (myFileTypes.length == 1) {
                myExportPlugin.fileType(myFileTypes[0]);
            } else {
                myExportPlugin.fileType(myFileTypes[i]);
            }

            myExportPlugin.performFunction(current);

        }

        fireProgress(100);
        return null;

    }

    /**
     * Button name for this plugin to be used in buttons, etc.
     *
     * @return String
     */
    @Override
    public String getButtonName() {
        return "Export Multiple";
    }

    /**
     * Tool Tip Text for this plugin
     *
     * @return String
     */
    @Override
    public String getToolTipText() {
        return "Export multiple data sets to files.";
    }

    public String[] getSaveFiles() {
        return mySaveFiles;
    }

    public void setSaveFiles(String[] saveFiles) {
        mySaveFiles = saveFiles;
    }

    public void setAlignmentFileTypes(FileLoadPlugin.TasselFileType[] types) {
        myFileTypes = types;
    }

    public void setAlignmentFileType(FileLoadPlugin.TasselFileType type) {
        myFileTypes = new FileLoadPlugin.TasselFileType[]{type};
    }

    public void setIncludeAnnotations(boolean include) {
        myExportPlugin.includeTaxaAnnotations(include);
    }

    public void setIncludeDepth(boolean include) {
        myExportPlugin.keepDepth(include);
    }

}

File: src/main/kotlin/net/maizegenetics/analysis/data/ExportMultiplePlugin.kt
/*
 * ExportMultiplePlugin.java
 *
 * Created on May 3, 2021
 *
 */
package net.maizegenetics.analysis.data

import net.maizegenetics.analysis.data.FileLoadPlugin.TasselFileType
import net.maizegenetics.plugindef.AbstractPlugin
import net.maizegenetics.plugindef.DataSet
import net.maizegenetics.plugindef.Datum
import net.maizegenetics.util.Utils
import org.apache.logging.log4j.LogManager
import java.io.File

/**
 * @author Terry Casstevens
 */
class ExportMultiplePlugin : AbstractPlugin(false) {

    private val logger = LogManager.getLogger(ExportMultiplePlugin::class.java)

    private val myExportPlugin = ExportPlugin()

    private var myFileTypes: Array<TasselFileType>? = null

    var saveFiles: Array<String>? = null

    override fun performFunction(input: DataSet): DataSet? {

        val data = input.dataSet
        var numSaveFiles = 0
        if (saveFiles != null) {
            numSaveFiles = saveFiles!!.size
        }
        check(!(numSaveFiles != 0 && numSaveFiles != 1 && numSaveFiles != data.size)) { "ExportMultiplePlugin: performFunction: number of save files should be either 0, 1 or number of input data sets." }
        if (myFileTypes != null && myFileTypes!!.size != 0) {
            check(!(myFileTypes!!.size != 1 && myFileTypes!!.size != data.size)) { "ExportMultiplePlugin: performFunction: number of files types should be either 0, 1 or number of input data sets." }
        }
        var i = 0
        val n = data.size
        while (i < n) {
            val datum = data[i] as Datum
            val current = DataSet(datum, input.creator)
            if (numSaveFiles == 0) {
                myExportPlugin.saveFile = datum.name
            } else if (numSaveFiles == 1) {
                var temp: String
                if (data.size == 1) {
                    temp = saveFiles!![0]
                } else {
                    val filename = Utils.getFilename(saveFiles!![0])
                    temp = filename.replaceFirst("\\.".toRegex(), (i + 1).toString() + ".")
                    if (temp.length == filename.length) {
                        temp = filename + (i + 1)
                    }
                    val directory = Utils.getDirectory(saveFiles!![0])
                    if (directory != ".") {
                        val dir = File(directory)
                        if (!dir.exists()) {
                            dir.mkdirs()
                        }
                        temp = "$directory/$temp"
                    }
                }
                myExportPlugin.saveFile = temp
            } else {
                myExportPlugin.saveFile = saveFiles!![i]
            }
            if (myFileTypes == null || myFileTypes!!.size == 0) {
                myExportPlugin.fileType = TasselFileType.Unknown
            } else if (myFileTypes!!.size == 1) {
                myExportPlugin.fileType = myFileTypes!![0]
            } else {
                myExportPlugin.fileType = myFileTypes!![i]
            }
            myExportPlugin.performFunction(current)
            i++
        }

        fireProgress(100)

        return null

    }

    /**
     * Button name for this plugin to be used in buttons, etc.
     *
     * @return String
     */
    override fun getButtonName(): String {
        return "Export Multiple"
    }

    /**
     * Tool Tip Text for this plugin
     *
     * @return String
     */
    override fun getToolTipText(): String {
        return "Export multiple data sets to files."
    }

    fun fileTypes(types: Array<TasselFileType>?) {
        myFileTypes = types
    }

    fun fileType(type: TasselFileType) {
        myFileTypes = arrayOf(type)
    }

    fun includeAnnotations(include: Boolean) {
        myExportPlugin.includeTaxaAnnotations = include
    }

    fun includeDepth(include: Boolean) {
        myExportPlugin.keepDepth = include
    }

}